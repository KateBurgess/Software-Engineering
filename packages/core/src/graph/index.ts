export * from './types';
export * from './nodes';
export * from './creation';
export * from './traversal';
