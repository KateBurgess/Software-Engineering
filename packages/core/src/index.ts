export * from './provenance-graph';

export const PROV_VERSION = 'av1';

export const sum = (a: number, b: number) => {
  return a + b;
};
