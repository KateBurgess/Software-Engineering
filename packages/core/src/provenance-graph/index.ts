export * from './types';
export * from './provenance-nodes';
export * from './provenance-creation';
