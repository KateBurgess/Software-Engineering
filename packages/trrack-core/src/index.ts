export default function Test() {
  return 'Test 2';
}
