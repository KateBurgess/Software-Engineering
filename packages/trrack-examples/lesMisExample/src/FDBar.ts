/* eslint-disable no-unused-vars */
/* eslint-disable class-methods-use-this */
import * as d3 from 'd3';

export default class Bars {
  private graph = undefined;

  private hoverOver: (id: string) => void;

  private hoverOut: () => void;

  private select: (id: string) => void;

  constructor(graph, hoverOver, hoverOut, select) {
    this.graph = graph;

    this.hoverOver = hoverOver;
    this.hoverOut = hoverOut;
    this.select = select;

    this.drawBar();
  }

  drawBar() {
    const nodeMap = new Map();

    for (let i = 0; i < this.graph.links.length; i += 1) {
      if (nodeMap.get(this.graph.links[i].source.id) === undefined) {
        nodeMap.set(this.graph.links[i].source.id, 1);
      } else {
        nodeMap.set(this.graph.links[i].source.id, nodeMap.get(this.graph.links[i].source.id) + 1);
      }
      if (nodeMap.get(this.graph.links[i].target.id) === undefined) {
        nodeMap.set(this.graph.links[i].target.id, 1);
      } else {
        nodeMap.set(this.graph.links[i].target.id, nodeMap.get(this.graph.links[i].target.id) + 1);
      }
    }

    const bars = d3.select('#bars');

    const width:number = +bars.attr('width') - 10;
    const height:number = +bars.attr('height') - 10;

    const edgePadding = 50;
    const bottomPadding = 90;
    const rightPadding = 4;
    const barPadding = 1;

    const g = bars.append('g');

    const scale = d3.scaleLinear();

    let keyList:string[] = Array.from(nodeMap.keys());
    const values:number[] = Array.from(nodeMap.values());

    keyList = keyList.sort((x, y) => {
      if (nodeMap.get(x) < nodeMap.get(y)) {
        return 1;
      } if (nodeMap.get(x) > nodeMap.get(y)) {
        return -1;
      }
      return 0;
    });

    scale.domain([0, Math.max(...values)]);
    scale.range([0, height - bottomPadding * 2]);

    // setting up the bar chart, as well as the bar chart on click and hover.
    g
      .selectAll('rect')
      .data(keyList)
      .enter()
      .append('rect')
      .attr('x', (d : any, i) => i * ((width - edgePadding) / keyList.length) + barPadding + edgePadding - rightPadding)
      .attr('y', (d : any) => height - scale(nodeMap.get(d)) - bottomPadding)
      .attr('width', (width - edgePadding) / keyList.length - barPadding)
      .attr('height', (d : any) => scale(nodeMap.get(d)))
      .attr('id', (d:any) => `${d}B`)
      .classed('bar', true)
      .on('click', (d) => this.select(d))
      .on('mouseover', (d) => this.hoverOver(d))
      .on('mouseout', (d) => this.hoverOut());

    const fakeScale = d3.scaleLinear();

    fakeScale.domain([Math.max(...values), 0]);
    fakeScale.range([0, height - bottomPadding * 2]);

    const yAxis = d3.axisLeft(fakeScale);

    const xScale = d3.scaleBand()
      .domain(keyList)
      .range([1 + edgePadding - rightPadding, 1 + width - rightPadding]);

    const xAxis = d3.axisBottom(xScale);
    // These are the axis
    g.append('g')
      .attr('class', 'axis')
      .attr('transform', `translate(${edgePadding - 6},${bottomPadding})`)
      .call(yAxis);

    // the y axis
    g.append('g')
      .attr('class', 'axis')
      .attr('transform', `translate(0,${height - bottomPadding + 1})`)
      .call(xAxis)
      .selectAll('text')
      .style('text-anchor', 'end')
      .attr('font-size', '11px')
      .attr('dx', '-1em')
      .attr('dy', '-.45em')
      .attr('transform', 'rotate(-90)');
    // The title
    g.append('text')
      .attr('transform', 'rotate(-90)')
      .attr('y', 0)
      .attr('x', 0 - (height / 2))
      .attr('dy', '1em')
      .style('text-anchor', 'middle')
      .text('Node Degree');
  }

  selectBar(id) {
    this.deselectAllBars();

    d3.select('#bars')
      .selectAll(`rect[id='${id}B']`)
      .classed('barSelected', true);
  }

  deselectAllBars() {
    d3.select('#bars')
      .selectAll('.barSelected')
      .classed('barSelected', false);
  }

  hoverBar(id) {
    d3.select('#bars')
      .selectAll(`rect[id='${id}B']`)
      .classed('barHover', true);
  }

  dehoverBars() {
    d3.select('#bars')
      .selectAll('.barHover')
      .classed('barHover', false);
  }
}
