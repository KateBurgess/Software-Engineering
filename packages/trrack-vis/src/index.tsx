import React, { FC, HTMLAttributes, ReactChild } from 'react';

import ProvVis from './components/ProvVis';
import { ProvVisCreator, UndoRedoButtonCreator } from './components/ProvVisCreator';
import { Config, EventConfig } from './Utils/EventConfig';

export interface Props extends HTMLAttributes<HTMLDivElement> {
  /** custom content, defaults to 'the snozzberries taste like snozzberries' */
  children?: ReactChild;
}

// Please do not use types off of a default export module or else Storybook Docs will suffer.
// see: https://github.com/storybookjs/storybook/issues/9556
/**
 * A custom Thing component. Neat!
 */
export const Thing: FC<Props> = ({ children }) => <div>{children || 'the snozzberries taste like snozzberries'}</div>;

export {
  ProvVis, EventConfig, Config, ProvVisCreator, UndoRedoButtonCreator,
};
