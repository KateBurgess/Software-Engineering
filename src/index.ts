// Import here Polyfills if needed. Recommended core-js (npm i -D core-js)
// import "core-js/fn/array.find"
// ...

import initProvenance from './Provenance/InitializeProvenance';
import { ProvenanceGraph } from './Interfaces/ProvenanceGraph';
import Provenance, { ActionFunction, SubscriberFunction } from './Interfaces/Provenance';
import {
  NodeMetadata,
  NodeID,
  Artifacts,
  Diff,
  isStateNode,
  ProvenanceNode,
  StateNode,
  RootNode,
  CurrentNode,
  Nodes
} from './Interfaces/NodeInterfaces';

export {
  initProvenance,
  ProvenanceGraph,
  Provenance,
  ActionFunction,
  SubscriberFunction,
  NodeMetadata,
  NodeID,
  Diff,
  RootNode,
  StateNode,
  ProvenanceNode,
  isStateNode,
  Nodes,
  CurrentNode,
  Artifacts
};
