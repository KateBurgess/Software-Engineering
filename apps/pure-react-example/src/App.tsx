import { Action, initializeActionFunctionRegistryTracker, initializeTrrack } from '@trrack/core';
import { useCallback, useMemo, useState } from 'react';
import Tree, { useTreeState } from 'react-hyper-tree';
import { TreeNode } from 'react-hyper-tree/dist/helpers/node';

function useBaseValue(initialValue: number) {
  const [baseValue, setBaseValue] = useState(initialValue);

  const addOneToBaseValue = useCallback(() => {
    setBaseValue((b) => b + 1);
  }, []);

  const removeOneFromBaseValue = useCallback(() => {
    setBaseValue((b) => b - 1);
  }, []);

  const action: Action = {
    registry_name: "add_one",
    label: "Add one to base value",
    do: {
      args: [],
    },
    undo: {
      args: [],
    },
  };

  return {
    baseValue,
    addOneToBaseValue,
    removeOneFromBaseValue,
    action,
  };
}

function App() {
  const {
    baseValue,
    action,
    addOneToBaseValue,
    removeOneFromBaseValue,
  } = useBaseValue(0);

  const trrack = useMemo(() => {
    const tracker = initializeActionFunctionRegistryTracker();
    tracker.register("add_one", addOneToBaseValue, removeOneFromBaseValue);
    return initializeTrrack(tracker);
  }, [addOneToBaseValue, removeOneFromBaseValue]);

  const { required, handlers } = useTreeState({
    data: trrack.tree,
    id: "test",
    defaultOpened: true,
  });

  open(required.data, trrack.graph.current);

  return (
    <div style={{ padding: "1em" }}>
      <h1>Action tracking</h1>
      <h4>{baseValue}</h4>
      <button onClick={() => trrack.applyAction(action)}>Add</button>
      <button onClick={() => trrack.undo()}>Undo</button>
      <button onClick={() => trrack.redo()}>Redo</button>
      <div style={{ margin: "1em" }}>
        <Tree
          {...required}
          {...handlers}
          gapMode="margin"
          setSelected={(node, isSelected) => {
            if (isSelected) trrack.goToNode(node.id);
          }}
        />
      </div>
    </div>
  );
}

export default App;

function open(nodes: TreeNode[], current: string) {
  nodes.forEach((node) => {
    node.setSelected(current === node.id);
    node.setOpened(true);

    if (node.children) open(node.children, current);
  });
}
