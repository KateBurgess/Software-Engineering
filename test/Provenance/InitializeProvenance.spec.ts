import { ActionFunction, initProvenance } from '../../src';

enum TodoStatus {
  DONE = 'DONE',
  ONGOING = 'ONGOING',
  FUTURE = 'FUTURE'
}

interface TodoItem {
  task: string;
  createdOn: string;
  status: TodoStatus;
  completedOn: string;
}

type Todos = TodoItem[];

interface ToDoListState {
  user: {
    name: string;
    totalTask: number;
  };
  todos: Todos;
}

function setupApp() {
  const state: ToDoListState = {
    user: {
      name: 'Kiran',
      totalTask: 1
    },
    todos: [
      {
        task: 'Add unit tests',
        createdOn: new Date().toISOString(),
        status: TodoStatus.ONGOING,
        completedOn: ''
      }
    ]
  };

  const provenance = initProvenance(state, false);

  provenance.addGlobalObserver(() => {
    console.log('Will always be called');
  });

  provenance.addObserver(['todos'], () => {
    console.log('Added Task');
  });

  provenance.addObserver(['user', 'totalTask'], (state?: ToDoListState) => {
    console.log('Now the total tasks are: ', state?.user.totalTask);
  });

  const addTask: ActionFunction<ToDoListState> = (state: ToDoListState, task: TodoItem) => {
    state.todos.push(task);
    state.user.totalTask = state.todos.length;
    return state;
  };

  return { state, provenance, addTask };
}

describe('Initialization', () => {
  const { provenance } = setupApp();
  it('provenance is a valid Provenance object', () => {
    const functionProperties = [
      'graph',
      'current',
      'root',
      'applyAction',
      'addObserver',
      'addGlobalObserver',
      'goToNode',
      'goBackNSteps',
      'goBackOneStep',
      'goForwardOneStep',
      'reset',
      'done',
      'exportState',
      'importState',
      'exportProvenanceGraph',
      'importProvenanceGraph'
    ];

    functionProperties.forEach(property => {
      expect(provenance).toHaveProperty(property);
    });
  });
});

describe('provenance.graph returns a valid provenance object', () => {
  const { provenance } = setupApp();

  it('has valid keys', () => {
    expect(provenance.graph()).toMatchSnapshot({
      current: expect.any(String),
      root: expect.any(String),
      nodes: expect.any(Object)
    });
  });

  const { nodes } = provenance.graph();

  it('has default task added', () => {
    expect(Object.keys(nodes).length).toBe(1);

    const defaultTaskId = Object.keys(nodes)[0];

    expect(nodes[defaultTaskId].state.todos.length).toBe(1);
  });

  test('default task matches the snapshot', () => {
    const defaultTaskId = Object.keys(nodes)[0];

    const defaultTask = nodes[defaultTaskId].state.todos[0];

    expect(defaultTask).toMatchInlineSnapshot(
      {
        task: 'Add unit tests',
        createdOn: expect.any(String),
        status: TodoStatus.ONGOING,
        completedOn: expect.any(String)
      },
      `
      Object {
        "completedOn": Any<String>,
        "createdOn": Any<String>,
        "status": "ONGOING",
        "task": "Add unit tests",
      }
    `
    );
  });
});

describe('applying an action', () => {
  const { provenance, addTask } = setupApp();

  const newTask: TodoItem = {
    createdOn: new Date().toISOString(),
    task: 'To check for coverage changes',
    status: TodoStatus.ONGOING,
    completedOn: ''
  };

  provenance.applyAction('Adding new task', addTask, [newTask]);

  test('New node has beend added with proper label', () => {
    expect(true).toBeTruthy();
  });
});
