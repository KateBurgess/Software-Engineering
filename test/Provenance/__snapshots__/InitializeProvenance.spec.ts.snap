// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provenance.graph returns a valid provenance object has valid keys 1`] = `
Object {
  "current": Any<String>,
  "nodes": Any<Object>,
  "root": Any<String>,
}
`;
